src/grlib.c

```diff
@@ -953,381 +951,659 @@ const struct grlib_ipcore gptimer = {
 /* New uart defines.  */
 #define UART_TX_TIME	1000
 #define UART_RX_TIME	1000
-#define UARTA_DR	0x1
-#define UARTA_SRE	0x2
-#define UARTA_HRE	0x4
-#define UARTA_OR	0x10
-
-/* UART support variables.  */
-
-#define APBUART0_IRQ 2
-#define APBUART1_IRQ 17
-
-/* File descriptor for input file.  */
-static int32 fd1, fd2;
+#define APBUART_STATUS_REG_OVERRUN	0x10
 
-/* UART status register */
-static int32 Ucontrol;
-
-static unsigned char aq[UARTBUF], bq[UARTBUF];
-static int32 anum, aind = 0;
-static int32 bnum, bind = 0;
-static char wbufa[UARTBUF], wbufb[UARTBUF];
-static unsigned wnuma;
-static unsigned wnumb;
-static FILE *f1in, *f1out;
-#ifdef HAVE_TERMIOS_H
-static struct termios ioc1, ioc2, iocold1, iocold2;
-#endif
 #ifndef O_NONBLOCK
 #define O_NONBLOCK 0
 #endif
 
-static int f1open = 0;
-
-static char uarta_sreg, uarta_hreg;
-static uint32 uart_stat_reg;
-static uint32 uarta_data;
-
-static int ifd1 = -1, ofd1 = -1;
+apbuart_type uarts[APBUART_NUM];
 
 void
 apbuart_init_stdio (void)
 {
-  if (dumbio)
-    return;
-#ifdef HAVE_TERMIOS_H
-  if (ifd1 == 0 && f1open)
+  uart_init_stdio(&uarts[0]);
+  uart_init_stdio(&uarts[1]);
+  uart_init_stdio(&uarts[2]);
+  uart_init_stdio(&uarts[3]);
+  uart_init_stdio(&uarts[4]);
+  uart_init_stdio(&uarts[5]);
+}
+
+void
+apbuart_restore_stdio (void)
+{
+  uart_restore_stdio(&uarts[0]);
+  uart_restore_stdio(&uarts[1]);
+  uart_restore_stdio(&uarts[2]);
+  uart_restore_stdio(&uarts[3]);
+  uart_restore_stdio(&uarts[4]);
+  uart_restore_stdio(&uarts[5]);
+}
+
+int
+uart_init_stdio(apbuart_type *uart)
+{
+  int result = 1;
+
+  if (uart != NULL) {
+    if (dumbio)
     {
-      tcsetattr (0, TCSANOW, &ioc1);
-      tcflush (ifd1, TCIFLUSH);
+      result = 0;
     }
+    else
+    {
+#ifdef HAVE_TERMIOS_H
+      if (uart->in_stream.descriptor == 0 && uart->device_open)
+      {
+        tcsetattr (0, TCSANOW, &uart->io_ctrl);
+        tcflush (uart->in_stream.descriptor, TCIFLUSH);
+        result = 0;
+      }
 #endif
+    }
+  }
+
+  return result;
 }
 
-void
-apbuart_restore_stdio (void)
+int
+uart_restore_stdio(apbuart_type *uart)
 {
-  if (dumbio)
-    return;
+  int result = 1;
+
+  if (uart != NULL)
+  {
+    if (dumbio)
+    {
+      result = 0;
+    }
+    else
+    {
 #ifdef HAVE_TERMIOS_H
-  if (ifd1 == 0 && f1open && tty_setup)
-    tcsetattr (0, TCSANOW, &iocold1);
+      if (uart->in_stream.descriptor == 0 && uart->device_open && tty_setup)
+      tcsetattr (0, TCSANOW, &uart->io_ctrl_old);
```

Comment:
Indent is misleading. Also, please add {} because it looks like a bug.

Reply:
ack